### PR TITLE
feat: middleware

### PR DESCRIPTION
**Pull request by Khaledgarbaya**
OK this is a very WIP and ugly first draft of the middleware feature.
I will need some help to change passing the middleware using a chaining function instead of an array.
For more context check https://github.com/yargs/yargs/issues/876

@bcoe I would love to get some feedback from you.

TODO : 
- [x]  Add documentation
- [x]  Add more tests
- [x]  Cleanup
- [x]  Rebase and squash commits

### CODE REVIEW

**Comment by bcoe** [comment]
@Khaledgarbaya thanks for opening this pull! I have a lot of OSS time this weekend and will give feedback.

**Comment by bcoe** [comment]
@Khaledgarbaya one thing that's been on my mind, before we extend yargs with too many more features, is I'd love to split the massive README.md in the root folder into a few docs in a `./docs` folder, I was thinking something along the lines of:

`docs/examples`.
`docs/basics`.
`docs/advanced-features`.
`docs/commands`.

Haven't fully thought it out yet -- but would like to perform this decomposition in the fairly near future as we build out some more features.

**Comment by Khaledgarbaya** [comment]
@bcoe I guess we can wait until the restructuring PR is done and we can discuss this PR. I have no problem with this.

**Comment by bcoe** [comment]
@Khaledgarbaya sorry 👋 have been pretty busy juggling various other works (working on some pretty cool things at npm 😄 ).

I've taken a stab at restructuring the README over the past week, and would love your feedback:

https://github.com/yargs/yargs/pull/892

What section would this middleware fall under do you think (maybe its own section in advanced topics?) or an addition to the command API?

My only real concern with the pull request currently, is that the `command()` method ends up taking a lot of arguments -- this is partially my fault for already having overloaded the method so much.

**Comment by Khaledgarbaya** [comment]
hi @bcoe I was thinking of that too recently, I am thinking of two ways. 

1. is to have a map at yargs level that links middlewares to a specific command.
e.g.

```js
var argv = require('yargs')
    .usage('Usage: $0 <command> [options]')
    .command('count', 'Count the lines in a file')
    .middlewares({'count': [middleware1, middlware2]})
    .argv;
```

2. or have the use `middlewares` of the command 
```js
var argv = require('yargs')
    .usage('Usage: $0 <command> [options]')
    .command('count', 'Count the lines in a file')
    .middlewares([middleware1, middlware2])
    .argv;
```
First option will let us to define middlewares that can run before each command for example if we specify `'*'`. Second option is not bad too but if we follow the current implementation of yargs `middlewares` should follow the same patterns as `aliases` etc .. meaning you should be able to pass it as property if you go the object route or `export` it inside the command module if you use `commandDir`.

In conclusion option 2 will still introduce a new argument to the command and option 1 will reduce the number of arguments by one and give us the possibility to run the same middleware for all the commands if we want.

Given the structure of yargs I am leaning tward solution 1 since it won't involve a lot of changes to the library and in the future maybe we can improve the command creation.

Best,
Khaled

**Comment by bcoe** [comment]
👋 hey @Khaledgarbaya sorry that I've been dropping this on the floor for so long, I've been on a couple trips this summer, and have also been dealing with some family issues

_long story short haven't had quite as many hours to work on OSS as I would like._

Now that I've refactored the docs for yargs, why don't you take a stab at adding a new advanced topic document on middleware, with a few practical examples; I think we can center the discussion of the optimal API around this document, and then work on getting this landed.

With yargs I've tried to take a slow methodical approach to evolving the API, which I understand can be a bit frustrating -- I promise we can get this landed though 👍

**Comment by Khaledgarbaya** [comment]
hey, @bcoe, no worries I was also on vacation and I will definitely try to find to move this PR forward.

File: lib/command.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ const inspect = require('util').inspect
 const camelCase = require('camelcase')
 
 const DEFAULT_MARKER = '*'
-
+const objectAssign = require('object-assign')
```
Comment (bcoe) [inline lib/command.js]: now that we require >=Node 4, you should be able to just use `Object.assign`.

**Comment by bcoe** [comment]
@Khaledgarbaya a good way to get this slick feature over the finish line would be to join the community slack that I recently created:

http://devtoolscommunity.herokuapp.com/

would love to coordinate with you, and help you land this.

**Comment by Khaledgarbaya** [comment]
Hey @bcoe, I made some changes and added the docs Sorry for taking so long

File: docs/advanced.md [hunk 2]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+## Midlewares
+Sometimes you want to perform operations on the args just before reaching the command.
+For example, you want to validate the args or check if the an argument is passed otherwise you load it from local file.
+Middlwares are basically functions ran before the command to prepare the args for them.
+
+Diagram:
+
+```
+                        --------------         --------------        ---------
+stdin ----> argv ----> | Middleware 1 | ----> | Middleware 2 | ---> | Command |
```
Comment (bcoe) [inline docs/advanced.md]: love this diagram 👍

File: docs/advanced.md [hunk 3]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+                    .option('password')
+      } ,(argv) => {
+        authenticateUser(argv.username, argv.password)
+      })
+  .middlewares([normalizeCredentials])
+  .argv;
+```
+
 
 See the [yargs-parser](https://github.com/yargs/yargs-parser#configuration) module
```
Comment (bcoe) [inline docs/advanced.md]: this section should be moved up to be with the "Customizing Yargs' Parser" section.

File: test/yargs.js [hunk 4]
```diff
@@ -301,7 +301,26 @@ describe('yargs dsl tests', () => {
         .exitProcess(false) // defaults to true.
         .argv
     })
-
+    it('run middlewares before reaching the handler', function (done) {
```
Comment (bcoe) [inline test/yargs.js]: slight grammar nit: "it('runs all middleware before reaching the handler'").

File: test/yargs.js [hunk 5]
```diff
@@ -301,7 +301,26 @@ describe('yargs dsl tests', () => {
...
+      yargs(['foo'])
+        .command(
+          'foo',
+          'handle foo things',
+          function () {},
+          function (argv) {
+            // we should get the argv filled with data from the middleware
+            argv._[0].should.equal('foo')
+            console.log(argv)
```
Comment (bcoe) [inline test/yargs.js]: we should get rid of this console.log.

File: test/yargs.js [hunk 6]
```diff
@@ -301,7 +301,26 @@ describe('yargs dsl tests', () => {
...
+            argv.hello.should.equal('world')
+            return done()
+          },
+        [function (argv) {
+          return {hello: 'world'}
```
Comment (bcoe) [inline test/yargs.js]: slick, so middleware is additive? we should definitely make sure we mention that it can both add new keys (while maintaining the existing) or overwrite old keys.

File: docs/advanced.md [hunk 7]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
   }
 }
 ```
+## Midlewares
```
Comment (bcoe) [inline docs/advanced.md]: Let's just make this heading: "Middleware".

File: docs/advanced.md [hunk 8]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+Sometimes you want to perform operations on the args just before reaching the command.
```
Comment (bcoe) [inline docs/advanced.md]: slight grammar nits:

```
Sometimes you might want to transform arguments before they reach the command handler.

For example, you perhaps you want to validate that credentials have been provided and otherwise load credentials from a file.

Middleware is simply a stack of functions, each of which is passed the the current parsed arguments, which it can in turn update by adding values, removing values, or overwriting values.
```

File: docs/advanced.md [hunk 9]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+Middlwares are basically functions ran before the command to prepare the args for them.
+
+Diagram:
+
+```
+                        --------------         --------------        ---------
+stdin ----> argv ----> | Middleware 1 | ----> | Middleware 2 | ---> | Command |
+                        --------------         --------------        ---------
+```
+### Usage
```
Comment (bcoe) [inline docs/advanced.md]: perhaps call this heading, "### Example Credentials Middleware"

File: docs/advanced.md [hunk 10]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+
+Here the middleware will check if the `username` and `password` is passed by the user
+if not it will load them from `~./credential` file and fil in the argv.username and argv.password.
```
Comment (bcoe) [inline docs/advanced.md]: slight grammar nits:

```
In this example, our middleware will check if the `username` and `password` is provided. If not, it will load them from `~/.credentials`, and fill in the `argv.username` and `argv.password` values.
```

File: docs/advanced.md [hunk 11]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+This means in your command you are sure that `argv.username` and `argv.password` have values in them.
```
Comment (bcoe) [inline docs/advanced.md]: you can probably drop this sentence (I think it's clear enough).

File: docs/advanced.md [hunk 12]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+const normalizeCredentials = (argv) => {
+  if (!argv.username || !argv.password) {
+    const credentials = JSON.parse(fs.readSync('~/.credentials'))
+    return credentials
+  }
+  return {}
+}
+```
+
+#### yarg
```
Comment (bcoe) [inline docs/advanced.md]: """#### yargs parsing configuration"""

File: docs/advanced.md [hunk 13]
```diff
@@ -403,6 +403,51 @@ some of yargs' parsing features:
...
+stdin ----> argv ----> | Middleware 1 | ----> | Middleware 2 | ---> | Command |
+                        --------------         --------------        ---------
+```
+### Usage
+
+Here the middleware will check if the `username` and `password` is passed by the user
+if not it will load them from `~./credential` file and fil in the argv.username and argv.password.
+This means in your command you are sure that `argv.username` and `argv.password` have values in them.
+
+#### middlware
```
Comment (bcoe) [inline docs/advanced.md]: "#### middleware function"

File: lib/command.js [hunk 14]
```diff
@@ -223,6 +223,12 @@ module.exports = function command (yargs, usage, validation) {
 
     if (commandHandler.handler && !yargs._hasOutput()) {
       yargs._setHasOutput()
+      if (commandHandler.middlewares.length > 0) {
+        const middlewareArgs = commandHandler.middlewares.reduce(function (initialObj, middleware) {
+          return Object.assign(initialObj, middleware(innerArgv))
```
Comment (bcoe) [inline lib/command.js]: to give people the ability to remove an argument in middleware, I think we should perhaps give middleware the actual object to modify, rather than modifying by reference with `Object.assign`, then a middleware function could:

`delete argv.password` and it would not show up in the final argv passed to the command. What do you think?

**Comment by bcoe** [comment]
@Morishiri interested in providing feedback on this feature? I'd love to unblock @Khaledgarbaya.

**Comment by Morishiri** [comment]
I just took a look at it and it looks really nice, interesting feature, really usefull :) Fully approved

**Comment by bcoe** [comment]
@Khaledgarbaya  thank you for the awesome contribution 🎉

**Comment by bcoe** [comment]
@Khaledgarbaya, @Darkylin please try `npm i yargs@next`, which has support for middleware. If everything is looking good I will promote the release.

**Comment by jacobrosenthal** [comment]
Im on yargs@next and Im attempting to use middleware as documented 
```
const normalizeCredentials = (argv) => {
  if (!argv.username || !argv.password) {
    const credentials = JSON.parse(fs.readSync('~/.credentials'))
    return credentials
  }
  return {}
}

var argv = require('yargs')
  .usage('Usage: $0 <command> [options]')
  .command('login', 'Authenticate user', (yargs) =>{
        return yargs.option('username')
                    .option('password')
      } ,(argv) => {
        authenticateUser(argv.username, argv.password)
      })
  .middlewares([normalizeCredentials])
  .argv;
```
but I get
```
bash-3.2$ node index.js --help
/Users/jacobrosenthal/Downloads/node-cryptokitties-cli/index.js:17
  .middlewares([normalizeCredentials])
   ^

TypeError: require(...).usage(...).command(...).middlewares is not a function
    at Object.<anonymous> (/Users/jacobrosenthal/Downloads/node-cryptokitties-cli/index.js:17:4)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:427:7)
    at startup (bootstrap_node.js:151:9)
    at bootstrap_node.js:542:3
bash-3.2$ 
```

Possible documentation or implementation error?

**Comment by bcoe** [comment]
@jacobrosenthal I believe this is an issue with documentation, @Khaledgarbaya ultimately decided that it made the most sense to pass middleware as the last argument to a `.command()`, rather than having a separate `middlewares` option.

@Khaledgarbaya, am I correct in this assertion? mind updating the docs when you have a moment?

**Comment by Khaledgarbaya** [comment]
Hey @bcoe @jacobrosenthal 
I am terribly sorry about that It totally slipped through.
I made a PR to fix that https://github.com/yargs/yargs/pull/1037

**Comment by jacobrosenthal** [comment]
That all worked great btw. However one follow up question. Is there any benefit of using the middleware argument vs calling a function (or two functions) in the command builder slot as far as internal lifecycle or something?

**Comment by bcoe** [comment]
@jacobrosenthal I think @Khaledgarbaya's idea was that you'd be able to apply a complex stack of transforms to all arguments in `argv` prior to it getting passed to the command handler...

You might be able to get similar behavior with a builder for individual arguments, whereas middleware processes the final argv object. @Khaledgarbaya out of curiosity do you have any real world examples of using this feature you can point us towards yet.
